Initial JWT cookie Authentication

gecage952 opened the pull request:
An initial implementation of cookie based auth with JWTs.

I'll put some context on session management below if anyone is interested, and why I think this approach works.

The basic idea of this to implement a JWT authentication flow. This flow works somewhat like this: After a user authenticates, some of their user info (really it's up to the developer's discretion as to which info) is encoded in a JWT which is signed with a secret that only the server knows and then returned to the user browser. This JWT can be stored as a cookie or in something like local storage. When making another request to the server (for navigation, etc), this JWT needs to be sent as part of the request. When using cookies, this JWT is sent automatically, and when using localstorage it needs to be attached in the auth header as a bearer token. When the server receives the request it verifies the token and gets the user info.  The pros of this approach is that you don't need to store user info anywhere on the server (stateless). The cons of this approach is that you typically don't want to save sensitive user info the JWT, considering that if someone were to grab the cookie (maybe by physically accessing the browser console), they could read the info in the JWT. The other con is that if you want to store a ton of user info, the cookie can get big storage-wise. 

This is opposed to session IDs. The Session ID flow (I'm just calling it that, I'm not sure what an "official" name would be) is when you have a database of session information in Redis or something similar with all of the keys being randomly generated strings. When a user authenticates, a new entry is created with that user's session info (can be whatever info you'd need) and the key is returned back to the user's browser as a cookie and called a Session ID. Whenever a user makes a request to the server, the cookie is sent along with the request. The server looks up the session ID in the database, attaches the user info to the request, does whatever it needs to do, and then returns the result to the user. The pros of this approach are that it is really easy to invalidate sessions from the server side, hence it being very secure. The cons of this approach is the overhead of storing user info in a database (albeit a fast one in Redis, etc).

In both approaches it's important that the cookies are secure, httponly, have samesite set, and have some expiration date. I think the JWT approach works well here, because we won't need to manage a database just for this one site to store sessions. We also aren't currently storing any sensitive info in our JWT cookie, as I'm using the ID token, not the access token for the cookie's contents. The one thing about saving any sort of token inside of the JWT cookie is that you are kind of double dipping when it comes to verification. Essentially, you verify the JWT cookie _and_ the ID/Access token on top of that. It is somewhat more secure, I guess, but not necessarily a ton. And the downside is that you just have that extra step to do, and if your identity provider is unavailable for some reason, the authentication would fail. From the user's perspective it's still super fast most of the time, but I suppose if enough people are using the system at once, it does add to the server's load. We can talk more in the next meeting if you want.

I'll also add that if we _would_ rather have the session ID flow, I could set that up instead. 


Some reading to help clarify: 
- https://github.com/OWASP/CheatSheetSeries/blob/master/cheatsheets/JSON_Web_Token_for_Java_Cheat_Sheet.md
- https://github.com/OWASP/CheatSheetSeries/blob/master/cheatsheets/Session_Management_Cheat_Sheet.md

gecage952 committed:
Implement initial keycloak authentication and cookie session management

gecage952 committed:
Use ID token instead of Access token for jwt

gecage952 committed:
Handle jwt decode errors more gracefully

marshallmcdonnell commented on intersect_registry_service/app/main.py:
```diff
@@ -75,6 +76,14 @@ async def lifespan(app: FastAPI) -> typing.AsyncGenerator[None, None]:
+app.add_middleware(
+    SessionMiddleware,
+    secret_key=settings.SESSION_SECRET,
+    max_age=60 * 60 * 24 * 7,
```
So am I reading this right as a 1 week max age?
Do we need to make that configurable or does it matter?

gecage952 replied:
We can make it configurable. Session can be pretty long for a lot of sites, although longer sessions typically employ the session ID approach detailed above. (By long, I mean like a couple of years).

marshallmcdonnell replied:
Cool! A default of 1 week is perfectly fine w/ me. But figured we can add it to something like `settings.SESSION_SECRET_MAX_AGE` maybe?

marshallmcdonnell commented on intersect_registry_service/app/core/environment.py:
```diff
@@ -78,6 +78,45 @@ class Settings(BaseSettings):
+    AUTHORIZE_URL: str = ''
+    """
+    Authentication URL From the OIDC provider.
+    """
```
Do we need to add an example of AUTHORIZE_URL to the `.env.example` file for the docker-compose yaml setup?
I was struggling a bit to set this up.

gecage952 replied:
Sorry, yeah, the .env should look something like this
```
SECRET_NAME="changemechangeme"
AUTH_IMPLEMENTATION=keycloak
AUTHORIZE_URL="http://127.0.0.1:8080/realms/DevRegistryKeycloak/protocol/openid-connect/auth"
TOKEN_URL="http://127.0.0.1:8080/realms/DevRegistryKeycloak/protocol/openid-connect/token"
JWKS_URL="http://127.0.0.1:8080/realms/DevRegistryKeycloak/protocol/openid-connect/certs"
SCOPE="openid email"
CLIENT_ID="registry-service-dev"
# CLIENT_ID="763661557184-1mv3jgajjoiub84fnto9q6v3jqdl6qna.apps.googleusercontent.com"
CLIENT_SECRET="secret"
# CLIENT_SECRET="secret"
REDIRECT_URL="http://localhost:8000/api/v1/auth/callback"

SESSION_SECRET=base64string (32/64 chars)
SESSION_FINGERPRINT_COOKIE="registry__Secure_Fgp"
```

gecage952 replied:
You will also need to go into the keycloak interface and manually import the configuration from a file (the json file provided). The auto import command is really finicky. I also changed the name of the Realm to remove the the spaces. I can show you in a call if that's easier.

marshallmcdonnell replied:
Thanks!
Sorry, do you mind to add this to the `.env.example` and maybe some instructions in the README?
Nothing major but just to capture these instructions.

gecage952 replied:
Sure

marshallmcdonnell reviewed (changes requested):
Overall, looking great!

I did have issues trying to get setup locally.
Maybe I just need to sync with @gecage952.
Mainly, I did need an AUTHORIZE_URL in the .env but do not think I'm adding the right URL (I tried http://localhost:8080; get to KeyCloak but just login, not the SSO w/ redirect)

Also, one small request for @gecage952 is to add yourself to the pyproject.toml authors list :)
Make sure to get credit!

Thanks for work on this!

Lance-Drane commented on pyproject.toml:
```diff
@@ -11,17 +11,21 @@ requires-python = ">=3.12,<4.0"
     "fastapi>=0.115.12",
     "fastapi-csrf-protect>=1.0.2",
     "fastapi-login>=1.10.3",
+    "httpx>=0.28.1",
```
Do we use httpx explicitly anywhere in the intersect_registry_service package? This is probably something we'd want as a dev dependency (especially once we add tests) but we shouldn't include it as required.

Lance-Drane commented on intersect_registry_service/app/core/environment.py:
```diff
@@ -78,6 +78,45 @@ class Settings(BaseSettings):
+    """
+    The name of the cookie that will be used to store the user fingerprint.
+    """
+
```
This is a note to myself - we _do_ want to have the empty string defaults, but we want to use [Pydantic Model Validator](https://docs.pydantic.dev/latest/concepts/validators/#model-validators) to check that these are not empty if `settings.AUTH_IMPLEMENTATION == 'keycloak'`.

Lance-Drane replied:
will add in separate commit

Lance-Drane commented on intersect_registry_service/app/auth/session.py:
```diff
@@ -18,15 +25,88 @@ def handle_unauthenticated(request: Request, exc: Exception) -> RedirectResponse
+    async def __call__(
+        self,
+        request: Request,
+    ) -> Any:
```
the return typing of this is always `USER`, right?

Lance-Drane commented on intersect_registry_service/app/auth/session.py:
```diff
@@ -18,15 +25,88 @@ def handle_unauthenticated(request: Request, exc: Exception) -> RedirectResponse
+                request,
+            )
+        except Exception:  # noqa: BLE001
+            return None
```
If this throws any exception other than `IntersectNotAuthenticatedError` , we should be logging it (probably as a warning, since I think that every exception which could be thrown requires the client to deliberately alter their cookie value).

(TBF I don't really see how any other Exception can be thrown, since `str.encode('utf-8')` shouldn't ever throw an exception in Python 3.)

Lance-Drane commented on intersect_registry_service/app/auth/session.py:
```diff
@@ -18,15 +25,88 @@ def handle_unauthenticated(request: Request, exc: Exception) -> RedirectResponse
+        Exception raised when no (valid) token is present.
+        Defaults to `fastapi_login.exceptions.InvalidCredentialsException`
+        """
+        return self._not_authenticated_exception
```
this can probably be removed (I wouldn't worry about implementing a custom exception to raise unless you plan on moving this code into a generic library)

Lance-Drane commented on intersect_registry_service/app/core/environment.py:
```diff
@@ -78,6 +78,45 @@ class Settings(BaseSettings):
+    REDIRECT_URL: str = ''
+    """
+    Should be /api/v1/auth/callback. The OIDC provider will redirect to this path after authentication.
+    """
```
It might make more sense for us to just have an environment variable which is just the _domain_ , and then we can use `functools.cached_property` to annotate a function which returns something like `f"${self.DOMAIN}${self.BASE_URL}/api/v1/auth/callback"` . The domain is something which a deployment engineer can easily customize, the path is not something which needs to be customizable at runtime beyond the base URL (which is really just the proxy portion of the path that's not generally used by the FastAPI application).

Lance-Drane replied:
fixed

Lance-Drane commented on Dockerfile.keycloak:
```diff
@@ -0,0 +1,16 @@
+# Replace this with real TLS certificates
+#ADD clientkeystore conf/server.keystore
+# RUN keytool -genkeypair -storepass password -storetype PKCS12 -keyalg RSA -keysize 2048 -dname "CN=server" -alias server -ext "SAN:c=DNS:localhost,IP:127.0.0.1" -keystore conf/server.keystore
+RUN /opt/keycloak/bin/kc.sh build
+
+ADD ${KEYCLOAK_REALM_PATH} /registry_keycloak_realm.json
+RUN /opt/keycloak/bin/kc.sh --verbose import --optimized --file /registry_keycloak_realm.json --override true
+
+
+ENTRYPOINT [ "/opt/keycloak/bin/kc.sh", "start-dev", "--db", "${KC_DB}", "--db-url", "${KC_DB_URL}", "--db-username", "${KC_DB_USERNAME}", "--db-password", "${KC_DB_PASSWORD}" ]
```
This doesn't need to be addressed now, but there's a few goals we'd probably like to have with our Keycloak deployments:

- allow people to be able to customize the realm.json file at runtime, and also provide a good base configuration for allowing the merger of JSON files (i.e. we could have a volume mount for json files and do some jq trickery to generate the file realms.json file)
- try to avoid having to maintain a Keycloak Docker image ourselves, this conflicts somewhat with the first bulletpoint but can be mitigated by having build files packaged with the Helm chart

this is something @marshallmcdonnell can help out with

marshallmcdonnell replied:
@gecage952 Thoughts on using the Bitnami Keycloak instead?
  - https://github.com/bitnami/containers/tree/main/bitnami/keycloak
  
  We don't need that for this PR.

But going forward, and moving to helm charts, any issues you saw using this container as "drop-in" replacement?

gecage952 replied:
That should be fine. We can just import the realm using the UI for keycloak. I actually think there might be an option to save all the realm info in the database too and have it automatically load it when configured properly.

marshallmcdonnell replied:
So from the helm chart repo, it looks like you can just mount in the JSON files for the realms:
  - link: https://github.com/bitnami/charts/blob/main/bitnami/keycloak/README.md#importing-a-realm

Does that work so you don't have to do it via the UI?

gecage952 replied:
Just being honest, the importer is very flakey. Like, it's definitely worth giving a shot, but I wouldn't be shocked if it gave us trouble haha :laughing:

marshallmcdonnell replied:
Ah, gotcha! Sorry if you already told me this, it rings a bell.

I did make this issue to "try" it out.
  - https://github.com/INTERSECT-SDK/registry-service/issues/4

I think after we get this merged in, I can take on that check.

Thank you!

Lance-Drane reviewed (commented):
This looks like it'll be a good start. I'll handle making the login approaches swappable, the comments scattered around the code are mostly just minor things I noticed when looking through the code.

gecage952 committed:
Add config option for setting session lifetime

gecage952 committed:
Expand realm config and document auth setup

gecage952 committed:
Remove google secret from realm config

marshallmcdonnell reviewed (approved):
Looks good for all my requested changes! Thank you!

I'd wait till @Lance-Drane is happy to merge but overall, great work!

gecage952 committed:
add option to turn off every-time id token verification

gecage952 commented:
Just added an option to turn off the every time verification of the ID token stored in the cookie. If the config is set to true, then it will check on every request (which means that if Keycloak is down all requests will be rejected essentially). If false, then it will continue to work even if keycloak is down for a period of time.

gecage952 commented:
Also as a last note, I did have my Google client secret in the repo. I don't really consider it to be sensitive since it's just a test client on Google anyways, but we can remove that from the git history if we really need to.

gecage952 commented on .dockerignore:
```diff
@@ -6,5 +6,6 @@
 !uv.lock
 !alembic.ini
 !migrations
+!dev_registry_keycloak_realm.json
```
Making a note, I actually need to update this

gecage952 committed:
Fix realm json file name in dockerignore

Lance-Drane committed:
finalize keycloak implementation, see commit for details

- make sure to remove Keycloak session when logging out

- make sure rudimentary auth approach works

- add a default user in the development realm config (still annoying that we have to go through the UI and create the realm...)

- reorganize the application in a way that the ugly auth implementation conditionals are confined to __init__.py files and that specific implementations never actually share a file

Signed-off-by: Lance-Drane <Lance-Drane@users.noreply.github.com>

Lance-Drane commented:
https://github.com/INTERSECT-SDK/registry-service/pull/3/commits/1d7225cc18bc59cba19f3acb622818d9b12c23e2 is a fairly large commit which makes a lot of changes, so @gecage952 @marshallmcdonnell you should probably make sure I didn't do anything outlandish, but when you all are happy with it we can go ahead and merge. As a summary of the changes I made:

- Add functionality to make sure that we log out of Keycloak _in addition_ to our own app when "Logout" is clicked (see https://www.keycloak.org/docs/latest/upgrading/index.html#openid-connect-logout for details on this)
- reorganize the project so that separate auth implementations and separate "development/production API modes" are in separate files. do the ugly conditionals in the __init__.py files instead of the files we actually read.
- added a new default user in the realm JSON config, you still have to import it through the UI but you don't have to bother with manually creating a new user unless you're testing out things like Google OAuth
- reorganize the environment variables so that there's fewer configuration values. As far as Keycloak URLs go, you just need to know the URL of the Keycloak realm and the domain of the registry service for the callback URL. Check .env.example or environment.py if you're curious what I did.

I moved the endpoints from `/api/v1/auth/*` to `/login/*` , to keep all of the auth implementations close together. I half considered getting rid of the "intermediate" login page on the Keycloak implementation entirely just to save users a page click, I kept it for now because 1) it gives a bit more context about what they're logging into and 2) it may be useful later for allowing multiple Keycloak-based auth mechanisms at once (the classic "sign in with Google/Github/etc."), though at the moment this doesn't matter.

marshallmcdonnell commented on intersect_registry_service/app/api/v1/endpoints/sdk/definitions.py:
```diff
@@ -0,0 +1,50 @@
+    """Response type used by INTERSECT-SDK Clients to understand how to connect to the INTERSECT ecosystem"""
+
+    client_name: str
+    """A randomly-generated temporary name, unique to your Client."""
```
So do we not reuse the SDK package as a dependency to define these?

Lance-Drane replied:
Yeah, we might want to make a package which has nothing but core definitions shared between various applications. I think this would be helpful for the registry service _response_ objects due to how complex they are; however, I don't think it's necessary for the _request_  params (for the useful annotations like `Query`, `Form`, and `Header`, it's also easier to separate each part out).

marshallmcdonnell reviewed (approved):
Overall, LGTM!

I only have one question but not blocking. 

I'm good with this to be merged!

gecage952 commented on intersect_registry_service/app/auth/impl_keycloak/get_user.py:
```diff
@@ -0,0 +1,28 @@
+from ..definitions import USER, IntersectNotAuthenticatedError
+
+
+def get_user(user_token: str) -> None | USER:
```
Personally, I would try to keep the optional verification here from the original keycloak file. I'm not sure that verifying this on every request in something we want in most cases, so it's nice to have the option.

Lance-Drane replied:
Optimized in https://github.com/INTERSECT-SDK/registry-service/pull/3/commits/e31067f80fc2157652f8232b38a99d2cdb858446 so that we cache this locally instead of calling Keycloak on every request. This also allows us to not bother with the `SESSION_VERIFY_ID` environment variable, we can always just assume that's "True" now.

gecage952 commented on intersect_registry_service/app/core/environment.py:
```diff
@@ -75,18 +97,44 @@ class Settings(BaseSettings):
     """
+
+    @cached_property
+    def keycloak_authorize_url(self) -> str:
```
we technically can pull these from the well known config uri. I think it would be doing a GET request to `http://keycloak-url/realms/DevRegistryKeycloak/.well-known/openid-configuration` and the response would just be a json with all of the endpoints defined. Up to you though.

Lance-Drane replied:
I'll take a look at it, I think this is a good idea because it looks like the actual endpoints can change without warning on a Keycloak update.

Lance-Drane replied:
I couldn't get it to work but would be interested in having this implemented later. Maybe in another MR.

gecage952 replied:
I can make an issue for it and we can save it for later if you want.

gecage952 commented on intersect_registry_service/app/ui/endpoints/login_keycloak.py:
```diff
@@ -0,0 +1,128 @@
+        settings.SESSION_FINGERPRINT_COOKIE,
+        secure=True,
+        httponly=True,
+        samesite='strict',
```
Does the samesite value have to be the same here as the one when it's set? Because it is `lax` above. (I have no idea, genuinely curious)

Lance-Drane replied:
Good catch, it looks like I do need to fix this.

`delete_cookie(name, *args)` is just a wrapper around

```python
response.set_cookie(name, max_age=0, expires=0, *args)
```

https://github.com/encode/starlette/issues/1207 gives better context on why Starlette added the `secure/httponly/samesite` parameters to `response.delete_cookie()`.

Lance-Drane committed:
optimize session id verification, remove DOMAIN and SESSION_VERIFY_ID env variables

the PyJWKClient handles caching of JWTs, so we can minimize calls to Keycloak

Signed-off-by: Lance-Drane <Lance-Drane@users.noreply.github.com>
